protocols/kad/src/query.rs

```diff
@@ -293,6 +335,91 @@ pub struct QueryResult<TInner, TPeers> {
     /// The opaque inner query state.
     pub inner: TInner,
     /// The successfully contacted peers.
-    pub peers: TPeers
+    pub peers: TPeers,
+    /// The collected query statistics.
+    pub stats: QueryStats
+}
+
+/// Execution statistics of a query.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct QueryStats {
+    requests: u32,
+    success: u32,
+    failure: u32,
+    start: Option<Instant>,
+    end: Option<Instant>
+}
+
+impl QueryStats {
+    pub fn empty() -> Self {
+        QueryStats {
+            requests: 0,
+            success: 0,
+            failure: 0,
+            start: None,
+            end: None,
+        }
+    }
+
+    /// Gets the total number of requests initiated by the query.
+    pub fn num_requests(&self) -> u32 {
+        self.requests
+    }
+
+    /// Gets the number of successful requests.
+    pub fn num_successes(&self) -> u32 {
+        self.success
+    }
+
+    /// Gets the number of failed requests.
+    pub fn num_failures(&self) -> u32 {
+        self.failure
+    }
+
+    /// Gets the number of pending requests.
+    ///
+    /// > **Note**: A query can finish while still having pending
+    /// > requests, if the termination conditions are already met.
+    pub fn pending(&self) -> u32 {
+        self.requests - (self.success + self.failure)
```

Comment:
```
thread 'main' panicked at 'attempt to subtract with overflow', /home/mxinden/.cargo/git/checkouts/rust-libp2p-800b9a
7842b0accc/ae27c37/protocols/kad/src/query.rs:413:9
```

```suggestion
        self.requests.checked_sub(self.success + self.failure).unwrap_or(0)
```

Reply:
Why would `success + failure` ever be larger than `requests`? At least that is not the intention, so there may be a bug somewhere.

Reply:
Can you provide more context for the panic?

Reply:
Sorry for the rather mechanical error report.

I have a guess what is happening: Failing to contact a node always notifies all `ClosestPeersIter`s. Say there are two `ClosestPeersIter`, both query some unavailable node. Thus each iterator `on_failure` will be called twice, increasing its `failures` stat twice instead of once.


By returning false on `PeerState` `Failed` and `Succeeded` one can ensure not to double count failures.

```diff
diff --git a/protocols/kad/src/query/peers/closest.rs b/protocols/kad/src/query/peers/closest.rs
index 78be44d2..a8f0fb13 100644
--- a/protocols/kad/src/query/peers/closest.rs
+++ b/protocols/kad/src/query/peers/closest.rs
@@ -235,8 +235,7 @@ impl ClosestPeersIter {
                 PeerState::Unresponsive => {
                     e.get_mut().state = PeerState::Failed
                 }
-                _ => {}
-            }
+                _ => return false,
         }
 
         true
```

Does this make sense @romanb?

Reply:
Yes, that looks like the bug, good catch.